orttraining/orttraining/python/training/optim/config.py

```diff
@@ -0,0 +1,195 @@
+
+    Example:
+
+    .. code-block:: python
+
+        lamb_optim = _OptimizerConfig(name = 'LambOptimizer',
+                                    hyper_parameters = {'lr': 0.001, 'alpha' : 0.01, 'beta' : 0.9},
+                                    param_groups = [ { 'params' : ['model_param_0', 'model_param1'],
+                                                       'epsilon' : 0.03, 'beta' : 0.5},
+                                                     { 'params' : ['model_param_2'],
+                                                       'alpha' : 0.04},
+                                                   ]
+                    )
+    """
+
+    def __init__(self, name, hyper_parameters, param_groups=[]):
+        assert isinstance(name, str), "'name' must be a string"
+        assert name in ['AdamOptimizer', 'LambOptimizer', 'SGDOptimizer'], \
+            "'name' must be one of 'AdamOptimizer', 'LambOptimizer' or 'SGDOptimizer'"
+        assert isinstance(hyper_parameters,
+                          dict), "'hyper_parameters' must be a dict"
+        assert 'lr' in hyper_parameters, "'hyper_parameters' must contain a {'lr' : positive number} entry"
+        assert (isinstance(hyper_parameters['lr'], float) or
+                isinstance(hyper_parameters['lr'], int)) and hyper_parameters['lr'] >= 0, "lr must be a positive number"
+        assert isinstance(param_groups, list), "'param_groups' must be a list"
+        for group in param_groups:
+            assert isinstance(group, dict) and len(group) > 1 and 'params' in group, \
+                ("Each dict inside 'param_groups' must contain a {'params' : [model parameter names]} entry"
+                 " and additional entries for custom hyper parameter values")
+            for k, _ in group.items():
+                if k != 'params':
+                    assert k in hyper_parameters, f"'param_groups' has 'k' hyper parameter not present at 'hyper_parameters'"
+
+        self.name = name
+        self.lr = hyper_parameters['lr']
+        self.hyper_parameters = hyper_parameters
+        self.param_groups = []
+
+        # TODO: monitor this for perf issues
+        # Maybe we don't have to do this to populate TrainingParameters,
+        # but it does make code easier to maintain
+        for param_group in param_groups:
+            self._add_param_group(param_group)
+
+    def _add_param_group(self, param_group):
+        r"""Add a parameter group to the :py:class:`_OptimizerConfig` s `param_groups`."""
+        assert isinstance(param_group, dict), "param group must be a dict"
+
+        # Each parameter group must have all hyper parameters set
+        for name, value in self.hyper_parameters.items():
+            if name not in param_group:
+                param_group.setdefault(name, value)
+
+        self.param_groups.append(param_group)
+
+
+class SGD(_OptimizerConfig):
```

Comment:
shall it be called SGDConfig? It is really configuration for the SGD optimizer. Same for other classes.

Reply:
Excellent point! There was a discussion with Derek and it was agreed that it shouldn't be called `optim.SGD`, but it could be `optim.SGDConfig` or `optim.config.SGD`, for example.

In practice, this is how the users would code both approaches

```python
from onnxruntime.capi.training import optim
sgd_config = optim.config.SGD()
```
or
```python
from onnxruntime.capi.training import optim
sgd_config = optim.SGDConfig()
```

Is there any preference? This PR implemented the former. @spandantiwari @BowenBao @rayankrish @mrry

Reply:
I have a slight preference for `optim.SGDConfig` rather than the nested namespace (to defend against ambiguity if the user does a wildcard import or renames the imported module).

Reply:
`XXXConfig` it is :)